### PR TITLE
Use vodozemac for the E2EE support

### PR DESCRIPTION
**Pull request by poljar**
This PR intends to replaces olm-rs with vodozemac. It's still in a messy state, but can already be tested. I have been using this for a couple of days in a custom weechat-matrix-rs build.

The missing things here are:
1. A new way to encrypt keys/values in  our store.
2. Migrations for the sled store.

One notable exception here is the backup support. Backup support requires support for `PkEncrypt` and `PkDecrypt` structs from libolm, the whole asymmetric encryption scheme is deemed to be insecure and has been abandoned in favor of [MSC3270].

In conclusion, if backup support is needed, libolm will still be used.

[MSC3270]: https://github.com/matrix-org/matrix-doc/pull/3270

### CODE REVIEW

**Comment by poljar** [comment]
I think this should be ready for review. Couple of things that weren't done as part of this, since it would blow up the PR:

1. Use the store-cipher to hash the keys in the indexedb crypto store
2. Use the store-cipher in the state stores

File: crates/matrix-sdk-sled/Cargo.toml [hunk 1]
```diff
@@ -27,6 +27,7 @@ futures-util = { version = "0.3.15", default-features = false }
 matrix-sdk-base = { path = "../matrix-sdk-base", features = ["store_key"] }
 matrix-sdk-common = { path = "../matrix-sdk-common" }
 matrix-sdk-crypto = { path = "../matrix-sdk-crypto", optional = true }
+matrix-sdk-store-encryption = { path = "../matrix-sdk-store-encryption" }
```
Comment (gnunicorn) [inline crates/matrix-sdk-sled/Cargo.toml:30]: am I missing something or where is this crate coming from?

File: crates/matrix-sdk-crypto/src/store/mod.rs [hunk 2]
```diff
@@ -54,12 +54,10 @@ use std::{
     sync::Arc,
 };
 
-use base64::DecodeError;
+#[cfg(feature = "indexeddb_cryptostore")]
+use indexed_db_futures::web_sys::DomException;
```
Comment (gnunicorn) [inline crates/matrix-sdk-crypto/src/store/mod.rs:58]: what is this needed for?

File: crates/matrix-sdk-crypto/src/olm/utility.rs [hunk 3]
```diff
@@ -75,11 +71,11 @@ impl Utility {
             signature.get(key_id.to_string()).ok_or(SignatureError::NoSignatureFound)?;
         let signature = signature.as_str().ok_or(SignatureError::NoSignatureFound)?;
 
-        let ret =
-            match self.inner.ed25519_verify(signing_key, &canonical_json, signature.to_owned()) {
-                Ok(_) => Ok(()),
-                Err(_) => Err(SignatureError::VerificationError),
-            };
+        let signature = vodozemac::Ed25519Signature::from_base64(signature).unwrap();
```
Comment (gnunicorn) [inline crates/matrix-sdk-crypto/src/olm/utility.rs:74]: I think that was remarked before already, but why don't we have `?` here again?

From the context it is not clear to me, why this might not just panic, if I just gave it a non base64-string here... Is there something I am missing? then please explain that in a `.expect()` (rather than an opaque `unwrap`)

**Review by gnunicorn — CHANGES_REQUESTED** [review]
I have only looked superficially over the crypto-part of things, I don't really know enough to give that a proper in-depth review. However, I've found a few remarks, I'd like to see addressed...

File: crates/matrix-sdk-sled/Cargo.toml (hunk 1 above)
Comment (gnunicorn) [reply crates/matrix-sdk-sled/Cargo.toml:30]: scratch that, didn't realize the other thing was merged under that name

File: crates/matrix-sdk-crypto/Cargo.toml [hunk 4]
```diff
@@ -51,12 +50,23 @@ anyhow = "1"
 # feature = testing only
 http = { version = "0.2.4", optional = true}
 
+[target.'cfg(not(target_arch = "wasm32"))'.dependencies.vodozemac]
+git = "https://github.com/matrix-org/vodozemac"
+rev = "b6e30ba43742e27f2c996f19f071a6c78e3b30ce"
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/Cargo.toml:55]: It is, but vodozemac may still have some breaking changes. Depending on a rev seems like the safer solution.

File: crates/matrix-sdk-crypto/src/olm/account.rs [hunk 5]
```diff
@@ -627,16 +581,15 @@ impl ReadOnlyAccount {
...
         // might have failed to upload them the last time this method was
         // called.
-        if self.one_time_keys().await.curve25519().is_empty() {
+        if self.one_time_keys().await.is_empty() {
             let count = self.uploaded_key_count();
             let max_keys = self.max_one_time_keys().await;
-            let max_on_server = (max_keys as u64) / 2;
 
-            if count >= max_on_server {
+            if count >= max_keys as u64 {
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/olm/account.rs:588]: The [vodozemac](https://matrix-org.github.io/vodozemac/vodozemac/olm/struct.Account.html#method.max_number_of_one_time_keys) docs explain this:

> Note: this differs from the libolm method of the same name, the libolm method returned the maximum amount of one-time keys the Account could hold and only half of those should be uploaded.

File: crates/matrix-sdk-crypto/src/olm/utility.rs (hunk 3 above)
Comment (poljar) [reply crates/matrix-sdk-crypto/src/olm/utility.rs]: Hmm, can't remember a previous remark, nor can I find one.

This is just an error, we should use `?`. Fixed by https://github.com/matrix-org/matrix-rust-sdk/pull/494/commits/5a1e2d0a60f99aa5ff8843be3b32a645158e35de.

File: crates/matrix-sdk-crypto/src/olm/account.rs [hunk 6]
```diff
@@ -286,40 +289,15 @@ impl Account {
...
-                            error =? e,
-                            "Found a matching Olm session yet decryption failed",
-                        );
-                        return Err(OlmError::SessionWedged(
-                            sender.to_owned(),
-                            sender_key.to_owned(),
-                        ));
-                    }
+                Err(_) => {
+                    continue;
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/olm/account.rs:300]: We should not warn about this, this will create a lot of false positives. You can have a bunch of `Session`s and you don't know which one should be used to decrypt a message.

So you go through the `Session`s by try and error, as soon as we have at least two `Session`s for a device we'll be spamming this error message needlessly.

We log actual decryption failures here: https://github.com/matrix-org/matrix-rust-sdk/blob/5a1e2d0a60f99aa5ff8843be3b32a645158e35de/crates/matrix-sdk-crypto/src/olm/account.rs#L329

File: crates/matrix-sdk-crypto/src/file_encryption/key_export.rs [hunk 7]
```diff
@@ -319,7 +321,8 @@ mod test {
         let encrypted = encrypt_key_export(&export, "1234", 1).unwrap();
         let decrypted = decrypt_key_export(Cursor::new(encrypted), "1234").unwrap();
 
-        assert_eq!(export, decrypted);
+        // TODO, what should we do about PartialEq here?
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/file_encryption/key_export.rs]: We used to rely on the fact that libolm exports a room key (or inbound group session) as a string, which we could compare using `PartialEq`.

Vodozemac exports instead a struct that wraps the string to disallow `PartialEq` between secret values. This in turn makes it harder to check if our export mechanism works correctly. We still can get to the actual string using `as_str()`, so doing that instead of adding a `PartialEq`. Fixed in: https://github.com/matrix-org/matrix-rust-sdk/pull/494/commits/a7aef38d384354af35d87de6e613d339ff7a5f89

File: crates/matrix-sdk-crypto/src/olm/account.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/olm/account.rs]: I added some clarification in https://github.com/matrix-org/matrix-rust-sdk/pull/494/commits/8e3587132f8acf65434bb31f6015ceb67c9af205.

File: crates/matrix-sdk-crypto/src/olm/account.rs [hunk 8]
```diff
@@ -286,40 +289,15 @@ impl Account {
...
-
-                if !matches {
-                    continue;
-                }
-            }
-
-            let ret = session.decrypt(message.clone()).await;
+            let ret = session.decrypt(message).await;
 
             match ret {
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/olm/account.rs]: Handled by https://github.com/matrix-org/matrix-rust-sdk/pull/494/commits/8e3587132f8acf65434bb31f6015ceb67c9af205 as well.

File: crates/matrix-sdk-crypto/src/store/mod.rs (hunk 2 above)
Comment (poljar) [reply crates/matrix-sdk-crypto/src/store/mod.rs]: It isn't. This PR was made before the big store changes happened, so this seems to be a leftover from a merge/rebase. Removed in https://github.com/matrix-org/matrix-rust-sdk/pull/494/commits/e7d0ac42ef8634fa8f79667b0dc3ef34df51ed30.

File: crates/matrix-sdk-crypto/src/file_encryption/attachments.rs [hunk 9]
```diff
@@ -22,7 +22,7 @@ use aes::{
     Aes256, Aes256Ctr,
 };
 use base64::DecodeError;
-use getrandom::getrandom;
+use rand::{thread_rng, RngCore};
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/file_encryption/attachments.rs:25]: I think so, vodozemac uses `thread_rng` and the `wasm-bindgen` based js bindings worked fine once the `js` feature for getrandom was enabled.

File: crates/matrix-sdk-crypto/src/olm/account.rs [hunk 10]
```diff
@@ -288,17 +288,17 @@ impl Account {
...
-                    continue;
-                }
+            if let Ok(p) = session.decrypt(message).await {
+                decrypted = Some((session.clone(), p));
+                break;
+            } else {
+                // An error here is completely normal, after all we don't know
+                // which session was used to encrypt a message. We will log a
+                // warning if no session was able to decrypt the message.
+                continue;
```
Comment (gnunicorn) [inline crates/matrix-sdk-crypto/src/olm/account.rs:301]: imho, not needed.
```suggestion
```

**Review by gnunicorn — APPROVED** [review]

File: crates/matrix-sdk-crypto/src/olm/account.rs (hunk 10 above)
Comment (poljar) [reply crates/matrix-sdk-crypto/src/olm/account.rs:301]: I'll leave this in, people might audit this crate sooner or later and they might have the same question you had.

File: crates/matrix-sdk-sled/src/cryptostore.rs [hunk 11]
```diff
@@ -480,96 +479,18 @@ impl SledStore {
...
-                    .map_err(|e| CryptoStoreError::Backend(anyhow!(e)))?;
-                self.tracked_users
-                    .insert(account.user_id().as_str(), &[true as u8])
-                    .map_err(|e| CryptoStoreError::Backend(anyhow!(e)))?;
-            }
+        if version <= 3 {
+            return Err(anyhow!(
+                "Unsupported database version, the database \
+                 format changed in a backwards incompatible way"
+            )
```
Comment (gnunicorn) [inline crates/matrix-sdk-sled/src/cryptostore.rs:492]: Should we maybe have an actual error-type for Upgrade-issues so that a higher level can react upon it by backing up and deleting the existing instance and restarting automatically? Right now, it could only show that message and can't assist the user in resolving the issue...

Comment (gnunicorn) [reply crates/matrix-sdk-sled/src/cryptostore.rs:492]: Dunno, depends on the error (e.g. `UnsupportedDatabaseVersion(current, latest)`) and the way we want to handle this. I can still foresee a scenario where you might have really old databases and don't want to enforce upgrades (e.g. anything from before this release or because too many upgrades have been skipped) or you want the surrounding lib decide upon the details of the migration, might even interact with the user to provide a path for backups or something...

Comment (poljar) [reply crates/matrix-sdk-sled/src/cryptostore.rs:492]: Alright, will add.

Comment (poljar) [reply crates/matrix-sdk-sled/src/cryptostore.rs]: Did so in https://github.com/matrix-org/matrix-rust-sdk/pull/494/commits/c43b1b9e9d3104d504b3336e593c42ae6605c429.

File: crates/matrix-sdk-crypto/src/olm/account.rs [hunk 12]
```diff
@@ -285,42 +288,17 @@ impl Account {
...
-                        ));
-                    }
-                }
+            if let Ok(p) = session.decrypt(message).await {
+                decrypted = Some((session.clone(), p));
+                break;
+            } else {
+                // An error here is completely normal, after all we don't know
+                // which session was used to encrypt a message. We will log a
+                // warning if no session was able to decrypt the message.
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/olm/account.rs:300]:
```suggestion
                // These errors are expected to happen since we don't know which
                // session was used to encrypt a message, so we will undoubtedly
                // attempt to decrypt using the wrong one. In the event that no
                // session manages to decrypt the message, we will log a warning
                // in upstream code.
```

My attempt to make it a bit more explicit.

File: crates/matrix-sdk-crypto/src/olm/account.rs [hunk 13]
```diff
@@ -655,15 +608,15 @@ impl ReadOnlyAccount {
...
 
         // If we have a known key count, check that we have more than
-        // max_one_time_Keys() / 2, otherwise tell the client to upload more.
+        // max_one_time_Keys(), otherwise tell the client to upload more.
         let max_keys = self.max_one_time_keys().await as u64;
-        // If there are more keys already uploaded than max_key / 2
-        // bail out returning false, this also avoids overflow.
-        if count > (max_keys / 2) {
+        // If there are more keys already uploaded than max_key bail out
+        // returning false, this also avoids overflow.
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/olm/account.rs:614]:
```suggestion
        // If there are more keys already uploaded than max_key, bail out
        // returning false. This also avoids overflow.
```

File: crates/matrix-sdk-crypto/src/olm/account.rs [hunk 14]
```diff
@@ -655,15 +608,15 @@ impl ReadOnlyAccount {
         let count = self.uploaded_key_count();
 
         // If we have a known key count, check that we have more than
-        // max_one_time_Keys() / 2, otherwise tell the client to upload more.
+        // max_one_time_Keys(), otherwise tell the client to upload more.
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/olm/account.rs:611]:
```suggestion
        // max_one_time_keys(), otherwise tell the client to upload more.
```

File: crates/matrix-sdk-crypto/src/olm/group_sessions/inbound.rs [hunk 15]
```diff
@@ -279,7 +276,7 @@ impl InboundGroupSession {
 
     /// Restore a Session from a previously pickled string.
     ///
-    /// Returns the restored group session or a `OlmGroupSessionError` if there
+    /// Returns the restored group session or a `UnpicklingError` if there
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/olm/group_sessions/inbound.rs:279]:
```suggestion
    /// Returns the restored group session or a `PickleError` if there
```

Probably a left over from the earlier name?

File: crates/matrix-sdk-crypto/src/olm/session.rs [hunk 16]
```diff
@@ -76,13 +72,13 @@ impl fmt::Debug for Session {
 impl Session {
     /// Decrypt the given Olm message.
     ///
-    /// Returns the decrypted plaintext or an `OlmSessionError` if decryption
+    /// Returns the decrypted plaintext or an `DecrypitonError` if decryption
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/olm/session.rs:75]:
```suggestion
    /// Returns the decrypted plaintext or an `DecryptionError` if decryption
```

File: crates/matrix-sdk-crypto/src/olm/signing/pk_signing.rs [hunk 17]
```diff
@@ -114,61 +102,38 @@ impl PickledSigning {
...
-            "keys": subkey.keys.clone(),
-            "usage": subkey.usage.clone(),
-        });
-
-        let message = serde_json::to_string(&subkey_without_signatures)
-            .expect("Can't serialize cross signing subkey");
-        let signature = self.sign(&message).await;
+        let json_subkey = serde_json::to_value(&subkey).expect("Can't serialize cross signing key");
+        let signature =
+            self.inner.sign_json(json_subkey).await.expect("Can't sign cross signing keys");
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/olm/signing/pk_signing.rs:136]:
```suggestion
            self.inner.sign_json(json_subkey).await.expect("Can't sign cross-signing keys");
```

File: crates/matrix-sdk-crypto/src/olm/signing/pk_signing.rs [hunk 18]
```diff
@@ -114,61 +102,38 @@ impl PickledSigning {
...
-        let subkey_without_signatures = json!({
-            "user_id": subkey.user_id.clone(),
-            "keys": subkey.keys.clone(),
-            "usage": subkey.usage.clone(),
-        });
-
-        let message = serde_json::to_string(&subkey_without_signatures)
-            .expect("Can't serialize cross signing subkey");
-        let signature = self.sign(&message).await;
+        let json_subkey = serde_json::to_value(&subkey).expect("Can't serialize cross signing key");
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/olm/signing/pk_signing.rs:134]:
```suggestion
        let json_subkey = serde_json::to_value(&subkey).expect("Can't serialize cross-signing key");
```

File: crates/matrix-sdk-crypto/src/olm/utility.rs [hunk 19]
```diff
@@ -125,13 +122,13 @@ mod test {
 
         let signing_key = "n469gw7zm+KW+JsFIJKnFVvCKU14HwQyocggcCIQgZY";
 
-        let utility = Utility::new();
+        let signing_key = Ed25519PublicKey::from_base64(signing_key)
+            .expect("The signing key wasn't proper base64");
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/olm/utility.rs:126]:
```suggestion
            .expect("The signing key is not valid base64");
```

File: crates/matrix-sdk-crypto/src/store/mod.rs [hunk 20]
```diff
@@ -645,25 +578,17 @@ pub enum CryptoStoreError {
...
-    /// The underlying Olm Account operation returned an error.
-    #[error(transparent)]
-    OlmAccount(#[from] OlmAccountError),
-
-    /// The underlying Olm session operation returned an error.
-    #[error(transparent)]
-    OlmSession(#[from] OlmSessionError),
+    /// Failed to decrypt an pickled object.
+    #[error("An object failed to be decrypted while unpickling")]
+    UnpicklingError,
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/store/mod.rs:583]:
```suggestion
    /// Failed to decrypt a pickled object.
    #[error("Decryption failed while unpickling object")]
    UnpicklingError,
```

File: crates/matrix-sdk-crypto/src/store/mod.rs [hunk 21]
```diff
@@ -645,25 +578,17 @@ pub enum CryptoStoreError {
...
 
-    /// The underlying Olm group session operation returned an error.
+    /// Failed to decrypt an pickled object.
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/store/mod.rs:585]:
```suggestion
    /// Failed to decrypt a pickled object.
```

File: crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs [hunk 22]
```diff
@@ -721,27 +752,33 @@ impl SasState<WeAccepted> {
...
+                last_event_time: Instant::now().into(),
+                started_from_request: self.started_from_request,
+                state: Arc::new(KeyReceived {
+                    sas: Mutex::new(established).into(),
+                    we_started: self.state.we_started,
+                    accepted_protocols: self.state.accepted_protocols.clone(),
+                }),
+            })
+        } else {
+            Err(self.cancel(true, CancelCode::UnexpectedMessage))
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs:780]: Is this right? This seems unrelated to the given message (i.e. `content`) and depends solely on the `self.inner.lock().unwrap().take()` call failing.

Perhaps the result of that call depends on an earlier message? But in any case, it's a bit unobvious what's happening.

File: crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs [hunk 23]
```diff
@@ -761,35 +798,39 @@ impl SasState<Accepted> {
...
+                    last_event_time: Instant::now().into(),
+                    started_from_request: self.started_from_request,
+                    state: Arc::new(KeyReceived {
+                        sas: Mutex::new(established).into(),
+                        we_started: true,
+                        accepted_protocols: self.state.accepted_protocols.clone(),
+                    }),
+                })
+            } else {
+                Err(self.cancel(true, CancelCode::UnexpectedMessage))
```
Comment (dkasak) [inline crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs:832]: Same question as above.

File: crates/matrix-sdk-sled/src/cryptostore.rs [hunk 24]
```diff
@@ -51,10 +51,18 @@ use tracing::debug;
...
 
-/// This needs to be 32 bytes long since AES-GCM requires it, otherwise we will
-/// panic once we try to pickle a Signing object.
-const DEFAULT_PICKLE: &str = "DEFAULT_PICKLE_PASSPHRASE_123456";
-const DATABASE_VERSION: u8 = 3;
+const DATABASE_VERSION: u8 = 4;
+
+// Table names that are used to derive a separate key for each tree. This ensure
+// that user ids encoded for different trees won't end up as the same byte
+// sequence. This prevents corelation attacks on our tree metadata.
```
Comment (dkasak) [inline crates/matrix-sdk-sled/src/cryptostore.rs:58]:
```suggestion
// that user IDs encoded for different trees won't end up as the same byte
// sequence. This prevents correlation attacks on our tree metadata.
```

**Review by dkasak — CHANGES_REQUESTED** [review]
Looks sensible overall. Left some minor style tweaks and a few questions/comments. After addressing those, it's good to go.

File: crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs (hunk 22 above)
Comment (poljar) [reply crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs:780]: We transition here from a Sas object into an `EstablishedSas` object using the `diffie_hellman()` method. This  method takes `self` since our private curve25519 key is ephemeral and can only be used once.

This in turn means that we can only transition into the `KeyReceived` state once, if we receive another `m.key.verification.key` message we can't proceed. We can either cancel here or ignore the message.

File: crates/matrix-sdk-crypto/src/identities/device.rs [hunk 25]
```diff
@@ -534,15 +535,14 @@ impl ReadOnlyDevice {
     }
 
     pub(crate) fn is_signed_by_device(&self, json: &mut Value) -> Result<(), SignatureError> {
-        let signing_key =
+        let key =
             self.get_key(DeviceKeyAlgorithm::Ed25519).ok_or(SignatureError::MissingSigningKey)?;
```
[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/identities/device.rs:539]: Good question. The inner type here comes from [ruma-common](https://docs.ruma.io/ruma_common/encryption/struct.DeviceKeys.html) so either Ruma needs to depend on vodozemac or we need to have a customized version of this type.

Now I do plan on introducing a customized version of this type, but there are still a couple of problems with deserializing the `BTreeMap<Box<DeviceKeyId>, String>,` into a `BTreeSet<DeviceKey>` where `DeviceKey` would be something like

```rust
enum DeviceKey {
     Ed25519(Ed25519PublicKey),
     Curve25519(Curve25519PublicKey),
     Other(String)
}
```

1. `DeviceKeyId` doesn't contain only the algorithm (`ed25519`, `curve25519`), it also contains the device id of the device that owns the key. This makes it hard to use the `DeviceKeyId` for the `Serde` tag.

2. `Serde` doesn't support the `Other` variant on non-unit enum variants nicely: https://github.com/serde-rs/serde/issues/1973.

Overall, maybe we'll end up with a customized type that will do this, though not immediately. It'll depend on how annoying the serialization/deserialization code will be.

File: crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs (hunk 22 above)
Comment (dkasak) [reply crates/matrix-sdk-crypto/src/verification/sas/sas_state.rs:780]: Right, so the fact that `self.inner` is `None` would ultimately indicate that we've received `m.key.verification.key` twice, so the error code is indeed justified. Thanks.

File: crates/matrix-sdk-crypto/src/identities/device.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (jplatte) [reply crates/matrix-sdk-crypto/src/identities/device.rs:539]: We have a bunch of types like that in Ruma. We should definitely have a tracking issue for these Vodozemac / Ruma integration points. I think a dependency either direction could be problematic and we'll more likely end up having a lower-level crate underneath both, but all of that can be discussed after this PR is merged 🙂

[the remark this answers is not in the document]
Comment (poljar) [reply crates/matrix-sdk-crypto/src/identities/device.rs:539]: That's actually what I had planned. There are a bunch of customized types in the crypto crate already, more will come soon. So the plan is to have some vodozemac powered types in a small crate that both the crypto crate and other weird e2ee related projects can use.
